superset-frontend/src/dashboard/components/nativeFilters/FilterConfigModal.tsx

```diff
@@ -0,0 +1,283 @@
+  config.map(filter => filter.id);
+
+/**
+ * This is the modal to configure all the dashboard-native filters.
+ * Manages modal-level state, such as what filters are in the list,
+ * and which filter is currently being edited.
+ *
+ * Calls the `save` callback with the new FilterConfiguration object
+ * when the user saves the filters.
+ */
+export function FilterConfigModal({
+  isOpen,
+  initialFilterId,
+  createNewOnOpen,
+  save,
+  onCancel,
+}: FilterConfigModalProps) {
+  const [form] = Form.useForm<NativeFiltersForm>();
+
+  const filterConfig = useFilterConfiguration();
+  const filterConfigMap = useFilterConfigMap();
+  // new filter ids may belong to filters that do not exist yet
+  const [newFilterIds, setNewFilterIds] = useState<string[]>([]);
+  // store ids of filters that have been removed but keep them around in the state
+  const [removedFilters, setRemovedFilters] = useState<Record<string, boolean>>(
+    {},
+  );
+  const filterIds = useMemo(
+    () => uniq([...getFilterIds(filterConfig), ...newFilterIds]),
+    [filterConfig, newFilterIds],
+  );
+  const getInitialCurrentFilterId = useCallback(
+    () => initialFilterId ?? filterIds[0],
+    [initialFilterId, filterIds],
+  );
+  const [currentFilterId, setCurrentFilterId] = useState(
+    getInitialCurrentFilterId,
+  );
+  // the form values are managed by the antd form, but we copy them to here
+  const [formValues, setFormValues] = useState<NativeFiltersForm>({
+    filters: {},
+  });
+  const wasOpen = usePrevious(isOpen);
+
+  const addFilter = useCallback(() => {
+    const newFilterId = generateFilterId();
+    setNewFilterIds([...newFilterIds, newFilterId]);
+    setCurrentFilterId(newFilterId);
+  }, [newFilterIds, setCurrentFilterId]);
+
+  useEffect(() => {
+    if (createNewOnOpen && isOpen && !wasOpen) {
+      addFilter();
+    }
+  }, [createNewOnOpen, isOpen, wasOpen, addFilter]);
+
+  const resetForm = useCallback(() => {
+    form.resetFields();
+    setNewFilterIds([]);
+    setCurrentFilterId(getInitialCurrentFilterId());
+    setRemovedFilters({});
+  }, [form, getInitialCurrentFilterId]);
+
+  function onTabEdit(filterId: string, action: 'add' | 'remove') {
+    if (action === 'remove') {
+      setRemovedFilters({
+        ...removedFilters,
+        // trash can button is actually a toggle
+        [filterId]: !removedFilters[filterId],
+      });
+      if (filterId === currentFilterId && !removedFilters[filterId]) {
+        // when a filter is removed, switch the view to a non-removed one
+        const lastNotRemoved = findLastIndex(
+          filterIds,
+          id => !removedFilters[id] && id !== filterId,
+        );
+        if (lastNotRemoved !== -1)
+          setCurrentFilterId(filterIds[lastNotRemoved]);
+      }
+    } else if (action === 'add') {
+      addFilter();
+    }
+  }
+
+  function getFilterTitle(id: string) {
+    return (
+      formValues.filters[id]?.name ?? filterConfigMap[id]?.name ?? 'New Filter'
```

Comment:
The `?` is strong with this line.
![riddler](https://user-images.githubusercontent.com/812905/102454743-57f1cf80-3ff3-11eb-8b6d-a6f2924b92e7.gif)